src/dag.rs

```diff
@@ -203,76 +144,122 @@ impl<G: GraphNode> Dag<G> {
-        // temp_stack as an intermediate.
-        if self.visit_stack.is_empty() {
-            let mut temp_stack = vec![self.unvisited_roots.pop()?];
-            while let Some(cur_idx) = temp_stack.pop() {
-                self.visit_stack.push(cur_idx);
-                for child_idx in &self.dag.edges[cur_idx] {
-                    temp_stack.push(*child_idx);
+    // Private method that advances the iterator while checking for cycles.
+    //
+    // This is the iterative version of the DFS-based toposort implementation
+    // (https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search).
+    // Basically: do a normal DFS starting from the sources, and a node is
+    // appended to the toposort as soon as it is closed.
+    //
+    // If cycles are present, `next_or_error()` will find one eventually and
+    // return a `DagError::Cycle`. Subsequent calls will return `None`.
+    fn next_or_error(&mut self) -> Result<Option<&'a G>, DagError<G::NodeId>> {
+        while let Some(v) = self.dfs_stack.pop() {
+            match self.node_state[v] {
+                NodeState::New => {
+                    self.node_state[v] = NodeState::Opened;
+                    // Don't actually pop yet.
+                    self.dfs_stack.push(v);
+                    // If any child of v is `Opened`, we fonud a cycle!
+                    // This includes the case of a self-loop at v.
+                    if self.dag.edges[v]
+                        .iter()
+                        .any(|&u| self.node_state[u] == NodeState::Opened)
+                    {
+                        // Some cleanup + subsequent calls will return `None`.
+                        self.dfs_stack.clear();
+                        return Err(DagError::Cycle(self.dag.nodes[v].id().borrow().clone()));
+                    }
+                    self.dfs_stack.extend(
+                        self.dag.edges[v]
+                            .iter()
+                            .copied()
+                            .filter(|&u| self.node_state[u] == NodeState::New),
+                    );
                 }
+                NodeState::Opened => {
+                    self.node_state[v] = NodeState::Closed;
+                    return Ok(Some(&self.dag.nodes[v]));
+                }
+                NodeState::Closed => {}
             }
         }
-        // Now we just cruise down the to_visit stack drinking a large bottle of
-        // cranberry juice listening to Fleetwood Mac.
-        Some(&self.dag.nodes[self.visit_stack.pop().unwrap()])
+        Ok(None)
     }
-}
 
-#[derive(Clone)]
-pub struct TopDown<'a, G: GraphNode> {
-    dag: &'a Dag<G>,
-    visit_stack: Vec<usize>,
-    // Note these "roots" don't have to be roots in the overall DAG, they are
-    // only roots of the top-down traversals we're doing.
-    unvisited_roots: Vec<usize>,
+    fn check_cycles(mut self) -> Option<DagError<G::NodeId>> {
```

Comment:
Ah - now I realise why you couldn't use `?` above! 

Instead of `Option<E>` you want `Result<(), E>` - that's the idiomatic way to say "doesn't return anything but can fail".